cache-filter/src/unique_callout.rs

```diff
@@ -0,0 +1,129 @@
+use crate::{info, warn};
+use proxy_wasm::{
+    hostcalls::{get_shared_data, set_shared_data},
+    types::Status,
+};
+use threescale::proxy::CacheKey;
+
+// Callout lock is acquired by placing a key-value pair inside shared data.
+// Since, only one thread is allowed to access shared data (host uses mutex) thus only one winner;
+pub fn set_callout_lock(
+    root_id: u32,
+    context_id: u32,
+    cache_key: &CacheKey,
+) -> Result<bool, Status> {
+    let request_key = format!("callout_{}", cache_key.as_string());
+    let root_id_str = root_id.to_string();
+    let root_id_str_bytes = root_id_str.as_bytes();
+    info!(
+        context_id,
+        "thread(id: {}): trying to set callout-lock for request(key: {})", root_id, request_key
+    );
+
+    // check if lock is already acquired or not
+    match get_shared_data(&request_key)? {
+        (_, None) => {
+            // lock not acquired yet, try to set lock.
+            // Note: Since two threads can get None for CAS at the same time, to avoid
+            // overwriting of lock by another thread, use any value for CAS other than 0.
```

Comment:
Missing info: what's the expected value of CAS on success? The reader does not have details at this point about these implicit behavior.

Reply:
I think logs also act as a way to build context around what is happening, right?